src/utils/reqwest_resume/mod.rs

```diff
@@ -0,0 +1,165 @@
+    }
+}
+
+/// A builder to construct the properties of a Request.
+///
+/// See [`reqwest::RequestBuilder`].
+#[derive(Debug)]
+pub struct RequestBuilder(reqwest::Client, reqwest::Method, reqwest::Url);
+impl RequestBuilder {
+    /// Constructs the Request and sends it the target URL, returning a Response.
+    ///
+    /// See [`reqwest::RequestBuilder::send()`].
+    pub fn send(&mut self) -> impl Future<Output = reqwest::Result<Response>> + Send {
+        let (client, method, url) = (self.0.clone(), self.1.clone(), self.2.clone());
+        async move {
+            let response = loop {
+                let builder = client.request(method.clone(), url.clone());
+                match builder.send().await {
+                    Err(err) if !err.is_builder() && !err.is_redirect() && !err.is_status() => {
+                        sleep(Duration::from_secs(1)).await
+                    }
+                    x => break x?,
+                }
+            };
+            let accept_byte_ranges =
+                if let Some(value) = response.headers().get(header::ACCEPT_RANGES) {
+                    value.as_bytes() == b"bytes"
+                } else {
+                    false
+                };
+            let resp = Response {
+                _client: client,
+                _method: method,
+                _url: url,
+                _response: response,
+                _accept_byte_ranges: accept_byte_ranges,
+                _pos: 0,
+            };
+            Ok(resp)
+        }
+    }
+}
+
+/// A Response to a submitted Request.
+///
+/// See [`reqwest::Response`].
+#[derive(Debug)]
+pub struct Response {
+    _client: reqwest::Client,
+    _method: reqwest::Method,
+    _url: reqwest::Url,
+    _response: reqwest::Response,
+    _accept_byte_ranges: bool,
+    _pos: u64,
+}
+
+impl Response {
+    pub fn response(self) -> reqwest::Response {
+        self._response
+    }
+}
+
+struct Decoder {
+    client: reqwest::Client,
+    method: reqwest::Method,
+    url: reqwest::Url,
+    decoder: Pin<Box<dyn Stream<Item = reqwest::Result<Bytes>> + Send + Unpin>>,
+    accept_byte_ranges: bool,
+    pos: u64,
+}
+impl Stream for Decoder {
+    type Item = reqwest::Result<Bytes>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        loop {
+            match ready!(self.decoder.as_mut().poll_next(cx)) {
+                Some(Err(err)) => {
+                    if !self.accept_byte_ranges {
+                        // TODO: we could try, for those servers that don't output Accept-Ranges but work anyway
+                        break Poll::Ready(Some(Err(err)));
+                    }
+                    let builder = self.client.request(self.method.clone(), self.url.clone());
+                    let mut headers = HeaderMap::new();
+                    let value = HeaderValue::from_str(&std::format!("bytes={}-", self.pos))
+                        .expect("invalid ASCII string");
```

Comment:
Could be changed to `unreachable`. Since `self.pos` is a `u64` - that should always work.

Reply:
What's the difference compared to a call to `expect`?

Reply:
I'm just talking about `expect("unreachable")`. I think it's more straightforward than whatever is there now.

Reply:
Ah ok, make sense. I thought you were talking about `unreachable!`.